src/wasm/wasm-binary.cpp

```diff
@@ -7962,17 +7999,79 @@ void WasmBinaryReader::visitResume(Resume* curr) {
   curr->finalize(&wasm);
 }
 
-void WasmBinaryReader::visitSuspend(Suspend* curr) {
+void WasmBinaryReader::visitResumeThrow(ResumeThrow* curr) {
```

Comment:
I think most (maybe all?) the shared logic concerns resume tables (i.e. `on` clauses). How about I abstract `handlerTags`, `onTags`, and `labels` into a new structure `ResumeTable`, which we attach to both `Resume` and `ResumeThrow`?

Reply:
The only problem with using a new kind of structure is that wasm-delegations-fields.def can really only handle Expression fields with very basic types. Because of that, it's probably better to keep the field definitions duplicated and just unify the logic that operates on them. We do this in a few places for the different call instructions, for instance.

Reply:
OK. If I template a, say, `ResumeExpression` class and let `Resume` and `ResumeThrow` subclass from it, then I can get wasm-delegations-fields.def to correctly map over the inherited (common) fields, right?

Reply:
I'm thinking more along the lines of having `visitResume` and `visitResumeThrow` call out to `template<typename T> void handleResumeClauses(T* curr) { ... }`.

Reply:
I've implemented your suggestion. I think we could push every into a templated function, if we allow ourselves to use some `constexpr` to check for `ResumeThrow` to handle extra field it has.

Reply:
I'm happy with how it is now. Better to abstract the right amount rather than over-extract and have to punch holes in the abstraction.